src/swarm-engine/Swarm/Game/Step.hs

```diff
@@ -174,15 +174,40 @@ type HasGameStepState sig m = (Has (State GameState) sig m, Has (Lift IO) sig m,
 --
 -- * Every tick, every active robot shall have exactly one opportunity to run.
 -- * The sequence in which robots are chosen to run is by increasing order of 'RID'.
-runRobotIDs :: HasGameStepState sig m => IS.IntSet -> m ()
-runRobotIDs robotNames = do
+runRobotIDs ::
+  HasGameStepState sig m =>
+  -- Optional RID to stop before
+  Maybe RID ->
+  IS.IntSet ->
+  m IS.IntSet
+runRobotIDs maybeStopBeforeRID robotNames = do
   time <- use $ temporal . ticks
-  flip (iterateRobots time) robotNames $ \rn -> do
+  flip (iterateRobots maybeStopBeforeRID time) robotNames $ \rn -> do
     mr <- uses (robotInfo . robotMap) (IM.lookup rn)
     forM_ mr (stepOneRobot rn)
  where
   stepOneRobot :: HasGameStepState sig m => RID -> Robot -> m ()
-  stepOneRobot rn rob = tickRobot rob >>= insertBackRobot rn
+  stepOneRobot rn rob = tickRobot rob >>= registerUpdatedRobotState rn
+
+-- | This function must be utilized in every loop that iterates over
+-- robots and invokes 'stepRobot' (i.e. steps a robot's CESK machine).
+getExtraRunnableRobots :: Has (State GameState) sig m => TickNumber -> m (IS.IntSet -> IS.IntSet)
+getExtraRunnableRobots time = do
```

Comment:
TODO: Add a comment about "guarding" the `union` operation with a `null` check; even though a "branchless" union may be semantically equivalent with shorter code, it would end up doing `O(log N)` operation ever time, rather than only when needed.